### DISCUSSION TITLE
Link Neighbor Loader explanation

### DISCUSSION THREAD

**Discussion by sciencewu** [pyg-team/pytorch_geometric#6826] [Q&A]
Hi, huge fan of PyG. Also of the discussions here, very helpful! 
I tried to join the Slack channel but it says there is an error.
My question is if I do this: `train_data = Batch.from_data_list(dataset[train_index])` and then `LinkNeighborLoader` will the negative edges be just within the graphs or with the other graphs?

I have added my code just to check what I might be doing wrong as the predictions are not as good:
```
# GNN model
class GAT(torch.nn.Module):
    def __init__(self, in_channels, hidden_channels, out_channels, heads, dropout, edge_dim):
        super().__init__()
        self.dropout = dropout

        self.conv1 = GATConv(in_channels = in_channels, 
                            out_channels = hidden_channels, 
                            heads = heads, 
                            edge_dim = edge_dim,
                            dropout=dropout
                            )
        self.conv2 = GATConv(in_channels = hidden_channels * heads, 
                            out_channels = out_channels,
                            edge_dim = edge_dim,
                            dropout=dropout
                            )

    def forward(self, x, edge_index, edge_attr):
        # x = F.dropout(x, p=self.dropout, training=self.training)
        x = F.elu(self.conv1(x, edge_index, edge_attr))
        # x = F.dropout(x, p=self.dropout, training=self.training)
        x = self.conv2(x, edge_index, edge_attr)
        return x

# Train-Test functions
def train(model, loader, optimizer):
    model.train()

    total_loss = 0
    for data in loader:
        data = data.to(device)
        optimizer.zero_grad()
        loss = F.cross_entropy(model(data.x, data.edge_index, data.edge_attr), data.y.long())
        total_loss += loss.item() * data.num_graphs
        loss.backward()
        optimizer.step()
    return total_loss / len(loader.dataset)

@torch.no_grad()
def encode(model, loader):
    model.eval()

    xs, ys = [], []
    for data in loader:
        data = data.to(device)
        out = model(data.x, data.edge_index, data.edge_attr)
        xs.append(out.argmax(dim=1))
        ys.append(data.y.cpu())
    return torch.cat(xs, dim=0).long().detach().cpu(), torch.cat(ys, dim=0).long().cpu()


# Test function
@torch.no_grad()
def test(model, train_loader, test_loader, fold):
    model.eval()

    x, y = encode(model, train_loader)
    pred = calculate_metrics(x.cpu(), y.cpu())

    metrics = {k+f"_{fold}_train":v for k,v in pred.items()}
    metrics['fold'] = fold
    metrics[f'fold_truths_{fold}_train'] = y.detach().cpu().numpy().tolist()
    metrics[f'fold_preds_{fold}_train'] = x.detach().cpu().numpy().tolist()

    x, y = encode(model, test_loader)
    
    pred = calculate_metrics(x.cpu(), y.cpu())

    metrics.update({k+f"_{fold}_test":v for k,v in pred.items()})
    metrics[f'fold_truths_{fold}_test'] = y.detach().cpu().numpy().tolist()
    metrics[f'fold_preds_{fold}_test'] = x.detach().cpu().numpy().tolist()

    return metrics

# Training procedure
kf = KFold(n_splits=5, random_state=0, shuffle=True)
output_dict = {}

for fold, (train_index, test_index) in tqdm(enumerate(kf.split(dataset))):

    # Group all training graphs into a single graph to perform sampling:
    train_data = Batch.from_data_list(dataset[train_index])
    sampler = ImbalancedSampler(train_data.y.long())
    loader = NeighborLoader(
                            train_data, 
                            batch_size=1024, 
                            #shuffle=True,
                            num_neighbors=[10, 10],
                            num_workers=1, 
                            persistent_workers=True,
                            sampler=sampler
                            )
    
    test_loader = DataLoader(dataset[test_index], batch_size=parameters['batch_size'])
    train_loader = DataLoader(dataset[train_index], batch_size=parameters['batch_size'])
    
    model = GAT(
        in_channels= dataset.num_features,
        out_channels=2,
        hidden_channels=parameters['hidden_dims'],
        edge_dim=dataset.num_edge_features,
        dropout=parameters['dropout'],
        heads=parameters['heads']
    ).to(device)
    optimizer = torch.optim.Adam(
                                 model.parameters(), 
                                 lr=parameters['learning_rate'], 
                                 weight_decay=parameters['weight_decay']
                                 )
    
    
    losses = []
    for epoch in tqdm(range(1, parameters['epochs'] + 1)):
        loss = train(model, loader, optimizer)
        losses.append(loss)

    
    # Results
    eval_ = test(model, train_loader, test_loader, fold)
    losses_dict = {f'losses_{fold}': losses}
    output_dict.update({**parameters,**eval_,  **losses_dict})
```

**Comment by rusty1s** [comment]
What problems do you have when trying to join the Slack?

`LinkNeighborLoader` acts like working on a single graph, so it will also create negatives between different graphs in that way. How big are your graphs though? Do you desperately need to make use of `LinkNeighborLoader` or can't you just iterate over your graphs in mini-batches (in case they are small).

**Reply by sciencewu** [reply]
Sorry, now it works. Until yesterday the slack link from PyG said that the link is not available anymore.

I have graphs of varying sizes (e.g. 10 and 300 nodes), and for each node of the graph i have a binary target. 
Currently I am basing it on the example `sage_unsup_ppi` example that you have offered based on the GraphSAGE paper. If there is a way to make it unsupervised without the `LinkNeighborLoader` I'm very interested to try.

The problem that I have is that if I use the node features independently from the graph the model has better predictive power than the GNN models. I can share the code here to see if I'm doing something wrong. The unsupervised model is performing better than the supervised model, where my guess is class imbalanceness (70-30), even after using the `ImbalancedSampler` for `NeighborLoader` in the supervised case.

**Reply by rusty1s** [reply]
Why do you necessarily want to use unsupervised learning if you already have labels?

If you still want to predict, I think this is more straightforward to do without link neighbor loader, e.g., for a given mini-batch, you remove a set of edges (e.g., via `dropout_edge`), sample negatives (e.g., via `batched_negative_sampling`), and then train with a binary loss against positive dropped edges and negative ones.

**Reply by sciencewu** [reply]
I would like to use supervised learning but using the model is only predicting the majority class due to class imbalanceness. 
Any clever way how to get around it?

Also any idea why only the node features have a better predictive power than when I add the topology?

**Reply by rusty1s** [reply]
I hear you, but still think that 70/30 is not really a high class imbalance, and if your model only outputs a single class, then I think there might be other issues.

**Reply by sciencewu** [reply]
I do not have access to the code right now, but once I have it would it be okay to go through it together to see if there is an issue?

**Reply by rusty1s** [reply]
Yeah, you can paste model architecture and training procedure here if that's possible.

**Reply by sciencewu** [reply]
Hi! I am sorry for the delay on the answer I had to reproduce the model from the scratch.
So the code that I am using is in the updated question now.
**My concern is why is the model not performing as good as only using the node features only (random forest)?** 
I am trying to predict a class for each node in the graph, but I'm training on several graphs.

**Reply by rusty1s** [reply]
Does the same hold if you use `SAGEConv` rather than `GATConv`?

**Reply by lisiq** [reply]
SAGEConv is performing better than GATConv here, it depends on the run but usually outperforms by 5-6%. 
Which is still ~20% worse than only using the node features as tabular data with random forest 🤔

**Reply by rusty1s** [reply]
Mh, that is weird. `SAGEConv` has a built-in skip-connection, which means that it theoretically could learn to not use any neighbors (resembling an MLP). It seems to struggle to do that, so it actually looks like your graph structure hurts the underlying model performance.

**Reply by lisiq** [reply]
Yes, that's what I was thinking as well. I will get a new batch of data which may be a bit different and if you are interested I will keep you noted:)

But [DeepGCN](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/ogbn_proteins_deepgcn.py) is performing way better, only ~2 % worse than random forest.
